File: tests/test_todo.py

import os
import unittest
from unittest.mock import patch, mock_open, call
from datetime import datetime
import sys
from io import StringIO
import subprocess # subprocessをモックするためにインポート

from scripts import todo

class TestTodoScript(unittest.TestCase):

    # テスト後に生成されたファイルをクリーンアップするためのセットアップ
    def setUp(self):
        self.test_dir = os.path.join(".gemini", "tasks")
        # Ensure the directory exists before starting tests that might need it
        os.makedirs(self.test_dir, exist_ok=True)
        self.todo_file_path = os.path.join(self.test_dir, "TODO-test-feature-branch.md")
        # テスト実行前にファイルが存在する場合は削除
        if os.path.exists(self.todo_file_path):
            os.remove(self.todo_file_path)

    # 各テストメソッドの後に実行されるクリーンアップ
    def tearDown(self):
        if os.path.exists(self.todo_file_path):
            os.remove(self.todo_file_path)
        # .gemini/tasks ディレクトリが空になったら削除
        # ただし、複数のテストが同じディレクトリを使う場合は注意が必要
        # ここでは、このテストスイートで作成されたファイルのみを削除する
        if os.path.exists(self.test_dir) and not os.listdir(self.test_dir):
            os.rmdir(self.test_dir)


    @patch('scripts.todo.get_branch_name')
    def test_init_creates_file_with_header(self, mock_get_branch_name):
        """
        `init` コマンドが正しいヘッダーを持つ新しいTODOファイルを作成することをテストします。
        """
        # --- 前準備 ---
        test_branch = "test-feature-branch"
        test_title = "My New Task"
        mock_get_branch_name.return_value = test_branch
        
        expected_path = os.path.join(".gemini", "tasks", f"TODO-{test_branch}.md")
        today_str = datetime.now().strftime('%Y-%m-%d')
        
        # --- 実行 ---
        todo.init(test_title)
        
        # --- 検証 ---
        # 1. ファイルが存在することを確認 (モックではなく実際のファイルシステムで)
        self.assertTrue(os.path.exists(expected_path))
        
        # 2. ファイルの内容を確認
        with open(expected_path, "r") as f:
            written_content = f.read()
        
        # ヘッダーの各部分が存在することを確認
        self.assertIn(f"# TASK: {test_title}", written_content)
        self.assertIn(f"- Branch: {test_branch}", written_content)
        self.assertIn(f"- Created: {today_str}", written_content)


    @patch('scripts.todo.get_branch_name')
    def test_add_appends_task_to_file(self, mock_get_branch_name):
        """
        `add` コマンドが新しいタスクをTODOファイルに追記することをテストします。
        """
        # --- 前準備 ---
        test_branch = "test-feature-branch"
        test_title = "Initial Task List"
        first_task = "Buy groceries"
        second_task = "Do laundry"
        mock_get_branch_name.return_value = test_branch

        # initでファイルを作成しておく
        todo.init(test_title)

        # --- 実行 ---
        todo.add(first_task)
        todo.add(second_task)

        # --- 検証 ---
        # ファイルの内容を確認
        with open(self.todo_file_path, "r") as f:
            content = f.read()
        
        # タスクが正しい形式で追記されていることを確認
        self.assertIn(f"- [ ] {first_task}\n", content)
        self.assertIn(f"- [ ] {second_task}\n", content)

        # タスクが既存のコンテンツの後に追記されていることを確認 (順序も含む)
        init_header_end_index = content.find(f"- Created: {datetime.now().strftime('%Y-%m-%d')}")
        first_task_index = content.find(f"- [ ] {first_task}")
        second_task_index = content.find(f"- [ ] {second_task}")

        self.assertGreater(first_task_index, init_header_end_index, "最初のタスクはヘッダーの後にあるべきです。")
        self.assertGreater(second_task_index, first_task_index, "2番目のタスクは1番目のタスクの後にあるべきです。")


    @patch('scripts.todo.get_branch_name')
    @patch('sys.stdout', new_callable=StringIO) # sys.stdoutをモック
    def test_show_displays_file_content(self, mock_stdout, mock_get_branch_name):
        """
        `show` コマンドがTODOファイルの内容を標準出力に表示することをテストします。
        """
        # --- 前準備 ---
        test_branch = "test-feature-branch"
        test_title = "Shopping List"
        task1 = "Apples"
        task2 = "Milk"
        mock_get_branch_name.return_value = test_branch

        # ファイルを初期化し、タスクを追加
        todo.init(test_title)
        todo.add(task1)
        todo.add(task2)

        # 期待される出力内容を構築 (ファイルの内容と一致するはず)
        expected_output_prefix = f"\n--- {os.path.basename(self.todo_file_path)} ---"
        with open(self.todo_file_path, "r") as f:
            expected_file_content = f.read().rstrip('\n') # show関数のrstrip('\n')に合わせる
        expected_full_output = expected_output_prefix + expected_file_content

        # --- 実行 ---
        todo.show()

        # --- 検証 ---
        # 標準出力の内容を確認
        self.assertEqual(expected_full_output, mock_stdout.getvalue())

    @patch('scripts.todo.get_branch_name')
    @patch('sys.stdout', new_callable=StringIO)
    def test_show_no_active_todo_message_if_file_not_exists(self, mock_stdout, mock_get_branch_name):
        """
        TODOファイルが存在しない場合に `show` コマンドが特定のメッセージを表示することをテストします。
        """
        # --- 前準備 ---
        test_branch = "test-feature-branch"
        mock_get_branch_name.return_value = test_branch
        
        # ファイルが存在しない状態を確実にする (setUpで削除されているはず)
        self.assertFalse(os.path.exists(self.todo_file_path))

        # --- 実行 ---
        todo.show()

        # --- 検証 ---
        # 標準出力の内容が期待されるメッセージであることを確認
        self.assertEqual("No active TODO for this branch.", mock_stdout.getvalue())


    @patch('scripts.todo.get_branch_name')
    def test_start_marks_first_matching_task_as_in_progress(self, mock_get_branch_name):
        """
        `start` コマンドが、最初に一致する未完了タスクを進行中 (`[/]`) に変更することをテストします。
        """
        # --- 前準備 ---
        test_branch = "test-feature-branch"
        mock_get_branch_name.return_value = test_branch
        todo.init("Tasks")
        todo.add("Task 1")
        todo.add("Task 2 [Important]")
        todo.add("Task 3")

        # --- 実行 ---
        todo.start("Task 2")

        # --- 検証 ---
        with open(self.todo_file_path, "r") as f:
            content = f.read()
        self.assertIn("- [/] Task 2 [Important]\n", content)
        self.assertIn("- [ ] Task 1\n", content) # 他のタスクは変更されない
        self.assertIn("- [ ] Task 3\n", content)

    @patch('scripts.todo.get_branch_name')
    @patch('sys.stdout', new_callable=StringIO)
    @patch('sys.exit')
    def test_start_exits_if_task_already_in_progress(self, mock_exit, mock_stdout, mock_get_branch_name):
        """
        既に進行中のタスクがある場合に `start` コマンドがエラーを表示して終了することをテストします。
        """
        # --- 前準備 ---
        test_branch = "test-feature-branch"
        mock_get_branch_name.return_value = test_branch
        todo.init("Tasks")
        todo.add("Task 1")
        todo.add("Task 2")
        todo.start("Task 1") # 最初のタスクを開始状態にする

        # --- 実行 ---
        todo.start("Task 2") # 別のタスクを開始しようとする

        # --- 検証 ---
        self.assertIn("ERROR: 他のタスクが実行中です。先に完了させてください。", mock_stdout.getvalue())
        mock_exit.assert_called_once_with(1)

    @patch('scripts.todo.get_branch_name')
    @patch('sys.stdout', new_callable=StringIO)
    @patch('sys.exit')
    def test_start_exits_if_no_matching_task_found(self, mock_exit, mock_stdout, mock_get_branch_name):
        """
        一致する未完了タスクが見つからない場合に `start` コマンドがエラーを表示して終了することをテストします。
        """
        # --- 前準備 ---
        test_branch = "test-feature-branch"
        mock_get_branch_name.return_value = test_branch
        todo.init("Tasks")
        todo.add("Task 1")

        # --- 実行 ---
        todo.start("NonExistent Task")

        # --- 検証 ---
        self.assertIn("Error: Task matching 'NonExistent Task' not found or already started.", mock_stdout.getvalue())
        mock_exit.assert_called_once_with(1)

    @patch('scripts.todo.get_branch_name')
    def test_done_marks_in_progress_task_as_completed(self, mock_get_branch_name):
        """
        `done` コマンドが現在進行中のタスク (`[/]`) を完了 (`[x]`) に変更することをテストします。
        """
        # --- 前準備 ---
        test_branch = "test-feature-branch"
        mock_get_branch_name.return_value = test_branch
        todo.init("Tasks")
        todo.add("Task A")
        todo.add("Task B")
        todo.start("Task A") # Task Aを進行状態にする

        # --- 実行 ---
        todo.done()

        # --- 検証 ---
        with open(self.todo_file_path, "r") as f:
            content = f.read()
        self.assertIn("- [x] Task A\n", content)
        self.assertIn("- [ ] Task B\n", content) # 他のタスクは変更されない

    @patch('scripts.todo.get_branch_name')
    @patch('sys.stdout', new_callable=StringIO)
    def test_done_no_in_progress_task_message_if_none(self, mock_stdout, mock_get_branch_name):
        """
        進行中のタスクがない場合に `done` コマンドが特定のメッセージを表示することをテストします。
        """
        # --- 前準備 ---
        test_branch = "test-feature-branch"
        mock_get_branch_name.return_value = test_branch
        todo.init("Tasks")
        todo.add("Task A")

        # --- 実行 ---
        todo.done()

        # --- 検証 ---
        self.assertEqual("No in-progress task found to mark as DONE.", mock_stdout.getvalue())


    @patch('sys.argv', ['todo.py'])
    @patch('sys.stdout', new_callable=StringIO)
    @patch('sys.exit')
    def test_main_exits_with_usage_if_no_args(self, mock_exit, mock_stdout):
        """
        引数なしで `main` を呼び出した場合、使用方法が表示され `sys.exit(1)` が呼び出されることをテストします。
        """
        # --- 実行 ---
        todo.main()

        # --- 検証 ---
        self.assertIn("Usage: todo.py [init|add|start|done|show] [args]", mock_stdout.getvalue())
        mock_exit.assert_called_once_with(1)

    @patch('sys.argv', ['todo.py', 'init', 'My Title'])
    @patch('scripts.todo.init')
    def test_main_calls_init_with_args(self, mock_init):
        """
        `main` が `init` コマンドと引数を適切に呼び出すことをテストします。
        """
        # --- 実行 ---
        todo.main()

        # --- 検証 ---
        mock_init.assert_called_once_with('My Title')

    @patch('sys.argv', ['todo.py', 'add', 'New Task'])
    @patch('scripts.todo.add')
    def test_main_calls_add_with_args(self, mock_add):
        """
        `main` が `add` コマンドと引数を適切に呼び出すことをテストします。
        """
        # --- 実行 ---
        todo.main()

        # --- 検証 ---
        mock_add.assert_called_once_with('New Task')

    @patch('sys.argv', ['todo.py', 'start', 'Task 1'])
    @patch('scripts.todo.start')
    def test_main_calls_start_with_args(self, mock_start):
        """
        `main` が `start` コマンドと引数を適切に呼び出すことをテストします。
        """
        # --- 実行 ---
        todo.main()

        # --- 検証 ---
        mock_start.assert_called_once_with('Task 1')

    @patch('sys.argv', ['todo.py', 'done'])
    @patch('scripts.todo.done')
    def test_main_calls_done(self, mock_done):
        """
        `main` が `done` コマンドを適切に呼び出すことをテストします。
        """
        # --- 実行 ---
        todo.main()

        # --- 検証 ---
        mock_done.assert_called_once()

    @patch('sys.argv', ['todo.py', 'show'])
    @patch('scripts.todo.show')
    def test_main_calls_show(self, mock_show):
        """
        `main` が `show` コマンドを適切に呼び出すことをテストします。
        """
        # --- 実行 ---
        todo.main()

        # --- 検証 ---
        mock_show.assert_called_once()

    @patch('sys.argv', ['todo.py', 'invalid_command'])
    @patch('sys.stdout', new_callable=StringIO)
    @patch('sys.exit')
    def test_main_exits_with_usage_if_invalid_command(self, mock_exit, mock_stdout):
        """
        不正なコマンドで `main` を呼び出した場合、使用方法が表示され `sys.exit(1)` が呼び出されることをテストします。
        """
        # --- 実行 ---
        todo.main()

        # --- 検証 ---
        self.assertIn("Usage: todo.py [init|add|start|done|show] [args]", mock_stdout.getvalue())
        mock_exit.assert_called_once_with(1)

    @patch('scripts.todo.subprocess.check_output')
    def test_get_branch_name_returns_correct_name(self, mock_check_output):
        """
        `get_branch_name` が実際の Git コマンドを実行し、正しいブランチ名を返すことをテストします。
        """
        # --- 前準備 ---
        mock_check_output.return_value = b'feature/my-new-feature\n'

        # --- 実行 ---
        branch_name = todo.get_branch_name()

        # --- 検証 ---
        mock_check_output.assert_called_once_with(["git", "rev-parse", "--abbrev-ref", "HEAD"])
        self.assertEqual("feature-my-new-feature", branch_name)

    @patch('scripts.todo.subprocess.check_output', side_effect=subprocess.CalledProcessError(1, "git"))
    def test_get_branch_name_returns_default_on_error(self, mock_check_output):
        """
        `get_branch_name` が Git コマンドでエラーが発生した場合に "default" を返すことをテストします。
        """
        # --- 実行 ---
        branch_name = todo.get_branch_name()

        # --- 検証 ---
        self.assertEqual("default", branch_name)

File: tests/test_todo.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import cp from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { mock } from 'node:test';

// scripts/todo.mjs を import する
import { getBranchName, init, add, show, start, done, main } from '../scripts/todo.mjs';

/**
 * テスト用の一時ディレクトリを作成し、環境変数を設定します。
 * 各テストの前後で呼び出すことで、テスト間の隔離を保証します。
 */
function setupTmpDir() {
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'todo-test-tasks-'));
  process.env.GEMINI_TASK_DIR = tmpDir;
  return tmpDir;
}

function teardownTmpDir(tmpDir) {
  if (tmpDir && fs.existsSync(tmpDir)) {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  }
}

test('todo.mjs コア機能', async (t) => {

  await t.test('ブランチ名の取得', async (t) => {
    await t.test('Git ブランチ名 feature/abc を feature-abc に変換する', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', (cmd, args) => {
        if (cmd === 'git' && args.includes('rev-parse')) {
          return { stdout: 'feature/abc\n', status: 0 };
        }
        return { status: 1 };
      });

      const branch = getBranchName();
      assert.strictEqual(branch, 'feature-abc');
      mockSpawnSync.mock.restore();
    });

    await t.test('Git エラー時は "default" を返す', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', () => {
        return { status: 1, error: new Error('git error') };
      });

      const branch = getBranchName();
      assert.strictEqual(branch, 'default');
      mockSpawnSync.mock.restore();
    });
  });

  await t.test('初期化', async (t) => {
    let tmpDir;
    t.beforeEach(() => {
      tmpDir = setupTmpDir();
    });
    t.afterEach(() => {
      teardownTmpDir(tmpDir);
    });

    await t.test('正しい形式でTODOファイルを初期化する', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', () => {
        return { stdout: 'test-branch\n', status: 0 };
      });

      const expectedPath = path.join(tmpDir, 'TODO-test-branch.md');
      
      init('Test Task');
      assert.ok(fs.existsSync(expectedPath), `ファイルが ${expectedPath} に作成される必要があります`);
      
      const content = fs.readFileSync(expectedPath, 'utf8');
      assert.ok(content.includes('# TASK: Test Task'));
      assert.ok(content.includes('- Branch: test-branch'));
      assert.ok(content.includes('- Created: '));
      
      mockSpawnSync.mock.restore();
    });
  });

  await t.test('追加', async (t) => {
    let tmpDir;
    t.beforeEach(() => {
      tmpDir = setupTmpDir();
    });
    t.afterEach(() => {
      teardownTmpDir(tmpDir);
    });

    await t.test('タスクを正しい形式で追記する', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', () => {
        return { stdout: 'test-branch\n', status: 0 };
      });
      const expectedPath = path.join(tmpDir, 'TODO-test-branch.md');
      
      init('Initial List');
      add('New Task');
      
      const content = fs.readFileSync(expectedPath, 'utf8');
      assert.ok(content.includes('- [ ] New Task\n'));
      
      mockSpawnSync.mock.restore();
    });
  });

  await t.test('表示', async (t) => {
    let tmpDir;
    t.beforeEach(() => {
      tmpDir = setupTmpDir();
    });
    t.afterEach(() => {
      teardownTmpDir(tmpDir);
    });

    await t.test('ファイル内容を標準出力に表示する', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', () => {
        return { stdout: 'test-branch\n', status: 0 };
      });
      const mockWrite = mock.method(process.stdout, 'write', () => {});
      const expectedPath = path.join(tmpDir, 'TODO-test-branch.md');

      init('Shopping');
      add('Apples');
      show();

      assert.ok(mockWrite.mock.callCount() >= 1);
      const lastCall = mockWrite.mock.calls[mockWrite.mock.callCount() - 1];
      assert.ok(lastCall.arguments[0].startsWith('\n--- TODO-test-branch.md ---# TASK: Shopping'));
      assert.ok(lastCall.arguments[0].includes('- [ ] Apples'));
      
      mockWrite.mock.restore();
      mockSpawnSync.mock.restore();
    });

    await t.test('ファイルが存在しない場合にメッセージを表示する', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', () => {
        return { stdout: 'non-existent-branch\n', status: 0 };
      });
      const mockWrite = mock.method(process.stdout, 'write', () => {});

      show();

      assert.ok(mockWrite.mock.callCount() >= 1);
      assert.strictEqual(mockWrite.mock.calls[0].arguments[0], 'No active TODO for this branch.');

      mockWrite.mock.restore();
      mockSpawnSync.mock.restore();
    });
  });

  await t.test('開始', async (t) => {
    let tmpDir;
    t.beforeEach(() => {
      tmpDir = setupTmpDir();
    });
    t.afterEach(() => {
      teardownTmpDir(tmpDir);
    });

    await t.test('指定したタスクを [ ] から [/] に変更する', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', () => {
        return { stdout: 'test-branch\n', status: 0 };
      });
      const expectedPath = path.join(tmpDir, 'TODO-test-branch.md');

      init('Test Task');
      add('Task 1');
      add('Task 2');
      
      start('Task 1');
      
      const content = fs.readFileSync(expectedPath, 'utf8');
      assert.ok(content.includes('- [/] Task 1\n'));
      assert.ok(content.includes('- [ ] Task 2\n'));
      
      mockSpawnSync.mock.restore();
    });

    await t.test('すでに [/] がある場合にエラーで終了する', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', () => {
        return { stdout: 'test-branch\n', status: 0 };
      });
      const mockExit = mock.method(process, 'exit', (code) => {
        throw new Error(`process.exit called with ${code}`);
      });
      const mockWrite = mock.method(process.stdout, 'write', () => {});

      init('Test Task');
      add('Task 1');
      start('Task 1');
      
      assert.throws(() => {
        start('Task 1');
      }, /process.exit called with 1/);

      assert.strictEqual(mockWrite.mock.calls[mockWrite.mock.callCount() - 1].arguments[0], 'ERROR: 他のタスクが実行中です。先に完了させてください。');

      mockExit.mock.restore();
      mockWrite.mock.restore();
      mockSpawnSync.mock.restore();
    });

    await t.test('start で一致するタスクが見つからない場合に、エラーメッセージを表示して process.exit(1) を呼ぶことを検証', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', () => {
        return { stdout: 'test-branch\n', status: 0 };
      });
      const mockExit = mock.method(process, 'exit', (code) => {
        throw new Error(`process.exit called with ${code}`);
      });
      const mockWrite = mock.method(process.stdout, 'write', () => {});

      init('Test Task');
      add('Task 1');
      
      assert.throws(() => {
        start('Non-existent Task');
      }, /process.exit called with 1/);

      assert.strictEqual(mockWrite.mock.calls[mockWrite.mock.callCount() - 1].arguments[0], "Error: Task matching 'Non-existent Task' not found or already started.");
      
      mockExit.mock.restore();
      mockWrite.mock.restore();
      mockSpawnSync.mock.restore();
    });

    await t.test('start 成功時に Started: <pattern> を標準出力に表示することを検証', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', () => {
        return { stdout: 'test-branch\n', status: 0 };
      });
      const mockWrite = mock.method(process.stdout, 'write', () => {});

      init('Test Task');
      add('Task 1');
      
      start('Task 1');
      
      assert.strictEqual(mockWrite.mock.calls[mockWrite.mock.callCount() - 1].arguments[0], 'Started: Task 1');
      
      mockWrite.mock.restore();
      mockSpawnSync.mock.restore();
    });
  });

  await t.test('完了', async (t) => {
    let tmpDir;
    t.beforeEach(() => {
      tmpDir = setupTmpDir();
    });
    t.afterEach(() => {
      teardownTmpDir(tmpDir);
    });

    await t.test('[/] のタスクを [x] に変更する', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', () => {
        return { stdout: 'test-branch\n', status: 0 };
      });
      const mockWrite = mock.method(process.stdout, 'write', () => {});
      const expectedPath = path.join(tmpDir, 'TODO-test-branch.md');

      init('Test Task');
      add('Task 1');
      start('Task 1');
      
      done();
      
      const content = fs.readFileSync(expectedPath, 'utf8');
      assert.ok(content.includes('- [x] Task 1\n'));
      
      mockWrite.mock.restore();
      mockSpawnSync.mock.restore();
    });

    await t.test('done 成功時に Task marked as DONE. を標準出力に表示することを検証', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', () => {
        return { stdout: 'test-branch\n', status: 0 };
      });
      const mockWrite = mock.method(process.stdout, 'write', () => {});

      init('Test Task');
      add('Task 1');
      start('Task 1');
      
      done();
      
      assert.strictEqual(mockWrite.mock.calls[mockWrite.mock.callCount() - 1].arguments[0], 'Task marked as DONE.');
      
      mockWrite.mock.restore();
      mockSpawnSync.mock.restore();
    });

    await t.test('進行中のタスクがない場合にメッセージを表示し、正常終了することを検証', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', () => {
        return { stdout: 'test-branch\n', status: 0 };
      });
      const mockWrite = mock.method(process.stdout, 'write', () => {});

      init('Test Task');
      add('Task 1');
      
      done();

      assert.strictEqual(mockWrite.mock.calls[mockWrite.mock.callCount() - 1].arguments[0], 'No in-progress task found to mark as DONE.');
      
      mockWrite.mock.restore();
      mockSpawnSync.mock.restore();
    });
  });

  await t.test('メインロジック', async (t) => {
    let tmpDir;
    t.beforeEach(() => {
      tmpDir = setupTmpDir();
    });
    t.afterEach(() => {
      teardownTmpDir(tmpDir);
    });

    await t.test('引数なしで Usage を出力し、終了コード 1 で終了する', (t) => {
      const mockExit = mock.method(process, 'exit', (code) => {
        throw new Error(`process.exit called with ${code}`);
      });
      const mockWrite = mock.method(process.stdout, 'write', () => {});

      assert.throws(() => {
        main(['node', 'todo.mjs']);
      }, /process.exit called with 1/);

      assert.strictEqual(mockWrite.mock.calls[0].arguments[0], 'Usage: todo.py [init|add|start|done|show] [args]');

      mockExit.mock.restore();
      mockWrite.mock.restore();
    });

    await t.test('init コマンドを正しく呼び出す', (t) => {
      const mockSpawnSync = mock.method(cp, 'spawnSync', () => {
        return { stdout: 'test-branch\n', status: 0 };
      });
      const expectedPath = path.join(tmpDir, 'TODO-test-branch.md');
      
      main(['node', 'todo.mjs', 'init', 'Main Test']);
      assert.ok(fs.existsSync(expectedPath));
      const content = fs.readFileSync(expectedPath, 'utf8');
      assert.ok(content.includes('# TASK: Main Test'));
      
      mockSpawnSync.mock.restore();
    });
  });

  await t.test('todo.mjs CLI 挙動', async (t) => {
    const scriptPath = path.resolve('scripts/todo.mjs');
    let tmpDir;
    t.beforeEach(() => {
      tmpDir = setupTmpDir();
    });
    t.afterEach(() => {
      teardownTmpDir(tmpDir);
    });

    await t.test('引数なしの場合に終了コード 1 と Usage を表示する', () => {
      const result = cp.spawnSync('node', [scriptPath], { 
        encoding: 'utf8',
        env: { ...process.env, GEMINI_TASK_DIR: tmpDir }
      });
      assert.strictEqual(result.status, 1);
      assert.strictEqual(result.stdout, 'Usage: todo.py [init|add|start|done|show] [args]');
    });

    await t.test('無効なコマンドの場合に終了コード 1 と Usage を表示する', () => {
      const result = cp.spawnSync('node', [scriptPath, 'invalid'], { 
        encoding: 'utf8',
        env: { ...process.env, GEMINI_TASK_DIR: tmpDir }
      });
      assert.strictEqual(result.status, 1);
      assert.strictEqual(result.stdout, 'Usage: todo.py [init|add|start|done|show] [args]');
    });

    await t.test('init 引数不足の場合に終了コード 1 と Usage を表示する', () => {
      const result = cp.spawnSync('node', [scriptPath, 'init'], { 
        encoding: 'utf8',
        env: { ...process.env, GEMINI_TASK_DIR: tmpDir }
      });
      assert.strictEqual(result.status, 1);
      assert.strictEqual(result.stdout, 'Usage: todo.py init <title>');
    });

    await t.test('add 引数不足の場合に終了コード 1 と Usage を表示する', () => {
      const result = cp.spawnSync('node', [scriptPath, 'add'], { 
        encoding: 'utf8',
        env: { ...process.env, GEMINI_TASK_DIR: tmpDir }
      });
      assert.strictEqual(result.status, 1);
      assert.strictEqual(result.stdout, 'Usage: todo.py add <task>');
    });

    await t.test('start 引数不足の場合に終了コード 1 と Usage を表示する', () => {
      const result = cp.spawnSync('node', [scriptPath, 'start'], { 
        encoding: 'utf8',
        env: { ...process.env, GEMINI_TASK_DIR: tmpDir }
      });
      assert.strictEqual(result.status, 1);
      assert.strictEqual(result.stdout, 'Usage: todo.py start <pattern>');
    });

    await t.test('CLI 経由で init, add, start, done, show が一連の動作として成功する', () => {
      const branchName = getBranchName();
      const expectedPath = path.join(tmpDir, `TODO-${branchName}.md`);
      
      // init
      let result = cp.spawnSync('node', [scriptPath, 'init', 'CLI Test'], { 
        encoding: 'utf8',
        env: { ...process.env, GEMINI_TASK_DIR: tmpDir }
      });
      assert.strictEqual(result.status, 0);
      assert.ok(fs.existsSync(expectedPath), `ファイルが ${expectedPath} に作成される必要があります`);

      // add
      result = cp.spawnSync('node', [scriptPath, 'add', 'Task from CLI'], { 
        encoding: 'utf8',
        env: { ...process.env, GEMINI_TASK_DIR: tmpDir }
      });
      assert.strictEqual(result.status, 0);

      // start
      result = cp.spawnSync('node', [scriptPath, 'start', 'Task from CLI'], { 
        encoding: 'utf8',
        env: { ...process.env, GEMINI_TASK_DIR: tmpDir }
      });
      assert.strictEqual(result.status, 0);
      assert.ok(result.stdout.includes('Started: Task from CLI'));

      // show
      result = cp.spawnSync('node', [scriptPath, 'show'], { 
        encoding: 'utf8',
        env: { ...process.env, GEMINI_TASK_DIR: tmpDir }
      });
      assert.strictEqual(result.status, 0);
      assert.ok(result.stdout.includes('Task from CLI'));
      assert.ok(result.stdout.includes('[/]'));

      // done
      result = cp.spawnSync('node', [scriptPath, 'done'], { 
        encoding: 'utf8',
        env: { ...process.env, GEMINI_TASK_DIR: tmpDir }
      });
      assert.strictEqual(result.status, 0);
      assert.ok(result.stdout.includes('Task marked as DONE.'));

      // verify final state
      const content = fs.readFileSync(expectedPath, 'utf8');
      assert.ok(content.includes('[x] Task from CLI'));
    });
  });
});
